src/main/java/duke/task/Task.java

```diff
@@ -0,0 +1,93 @@
+package duke.task;
+
+import duke.exception.DukeException;
+
+/**
+ * Abstract class representing a task.
+ *
+ * @author Lian Kok Hai
+ */
+public abstract class Task {
+    protected String taskName;
+    protected String type;
+    protected boolean isDone;
+
+    /**
+     * Constructs a task with given name.
+     *
+     * @param taskName
+     */
+    public Task(String taskName) {
+        this.taskName = taskName;
+        this.isDone = false;
+    }
+
+    /**
+     * Marks task as done.
+     */
+    public void markDone() {
+        this.isDone = true;
+    }
+
+    /**
+     * Marks task as undone.
+     */
+    public void unmarkDone() {
+        this.isDone = false;
+    }
+
+    public String getTaskName() {
+        return this.taskName;
+    }
+
+    /**
+     * Encodes a task into a String representation to save into text file.
+     * Implementation within each specific class.
+     *
+     * @return Encoded String representation of task.
+     */
+    public abstract String encode();
+
+    /**
+     * Decodes a String representation of a task into a Task object.
+     *
+     * @param str Encoded String representation of task.
+     * @return Decoded Task object.
+     * @throws DukeException Thrown when invalid String representation given.
+     */
+    public static Task decode(String str) throws DukeException {
+        String[] splitStr = str.split(" \\| ", 5);
+        Task result = null;
+        switch(splitStr[0]) {
```

Comment:
I like how the switch statements has the correct form as it is a common mistake.